python/rikai/spark/functions/vision.py

```diff
@@ -206,6 +215,152 @@ def video_to_images(
+                ]
+            ),
+            True,
+        ),
+        StructField(
+            "data",
+            StructType(
+                [
+                    StructField("width", IntegerType(), True),
+                    StructField("height", IntegerType(), True),
+                    StructField("num_frames", IntegerType(), True),
+                    StructField("duration", FloatType(), True),
+                    StructField("bit_rate", IntegerType(), True),
+                    StructField("frame_rate", IntegerType(), True),
+                    StructField("codec", StringType(), True),
+                    StructField("size", IntegerType(), True),
+                ]
+            ),
+            True,
+        ),
+    ]
+)
+
+
+@udf(returnType=video_metadata_schema)
+def video_metadata(video: Union[str, VideoStream]) -> dict:
+    """Return useful video stream metadata about the given video
+
+    Parameters
+    ----------
+    video: str or VideoStream
+        The video uri or the Video object
+
+    Returns
+    -------
+    result: dict
+        { error: dict error info if ffprobe threw; otherwise None,
+          data: { width, height, num_frames, duration, bit_rate, frame_rate,
+                  codec, size}}
+
+    Notes
+    -----
+    The frame_rate is rounded to the nearest whole number of frames per sec
+
+    Examples
+    --------
+    The following returns the fps rounded to the nearest integer
+    ```
+    import rikai.spark.functions as RF
+    (spark.createDataFrame([(VideoStream(<uri>),)], ['video'])
+          .withColumn('meta', RF.video_metadata('video'))
+          .select('meta.data.frame_rate'))
+    ```
+    """
+    probe_result = _probe(video)
+    if probe_result["error"] is not None:
+        return probe_result
+
+    video_stream = next(
+        (
+            stream
+            for stream in probe_result["data"].get("streams", [])
+            if stream.get("codec_type", None) == "video"
+        ),
+        None,
+    )
+    if video_stream is None:
+        return _error("No video stream found", probe=probe_result)
+
+    try:
+        return _data(
+            {
+                "width": _int_or_none(video_stream, "width"),
+                "height": _int_or_none(video_stream, "height"),
+                "num_frames": _int_or_none(video_stream, "nb_frames"),
+                "frame_rate": _fps_or_none(video_stream),
+                "duration": _float_or_none(video_stream, "duration"),
+                "bit_rate": _int_or_none(video_stream, "bit_rate"),
+                "codec": video_stream.get("codec_name", None),
+                "size": _int_or_none(
+                    probe_result["data"].get("format", {}), "size"
+                ),
+            }
+        )
+    except Exception as e:
+        return _error(str(e))
+
+
+def _probe(video):
+    try:
+        import ffmpeg
+    except ImportError:
+        raise ValueError(
```

Comment:
why not re-raise `ImportError` like

```raise ImportError(...) from e```?

Reply:
copy-paste